src/Compiler/Checking/TailCallChecks.fs

```diff
@@ -0,0 +1,868 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+module internal FSharp.Compiler.TailCallChecks
+
+open Internal.Utilities.Collections
+open Internal.Utilities.Library
+open Internal.Utilities.Library.Extras
+open FSharp.Compiler
+open FSharp.Compiler.AbstractIL.IL
+open FSharp.Compiler.DiagnosticsLogger
+open FSharp.Compiler.Features
+open FSharp.Compiler.Syntax
+open FSharp.Compiler.TcGlobals
+open FSharp.Compiler.TypedTree
+open FSharp.Compiler.TypedTreeBasics
+open FSharp.Compiler.TypedTreeOps
+open FSharp.Compiler.TypeRelations
+
+let PostInferenceChecksStackGuardDepth = GetEnvInteger "FSHARP_TailCallChecks" 50
+
+//--------------------------------------------------------------------------
+// check environment
+//--------------------------------------------------------------------------
+
+type env =
+    {
+        /// Values in module that have been marked [<TailCall>]
+        mustTailCall: Zset<Val>
+
+        /// Recursive expressions of [<TailCall>] attributed values
+        mustTailCallExprs: Map<Stamp, Expr>
+    }
+
+    override _.ToString() = "<env>"
+
+let (|ValUseAtApp|_|) e =
+    match e with
+    | InnerExprPat (Expr.App(funcExpr = InnerExprPat (Expr.Val (valRef = vref; flags = valUseFlags))) | Expr.Val (valRef = vref
+                                                                                                                  flags = valUseFlags)) ->
+        Some(vref, valUseFlags)
+    | _ -> None
+
+type IsTailCall =
```

Comment:
How about just `TailCall` as the name of the type? Things like `YesFromVal` on top of `IsTailCall` sound a bit odd to me, although that might be a personal pref.

Reply:
You're right. I changed the name.